perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -3030,22 +3045,104 @@ private void categorizeMembersForLightweightSynchronization(PerunSession sess, G
+	 * @return richMember or null
+	 */
+	private RichMember processRichUserExtSourceDuringCategorization(PerunSession sess, RichUserExtSource key, Map<UserExtSource, RichMember> mappingStructure) {
+		//check if ues is in the map
+		if (mappingStructure.containsKey(key.asUserExtSource())) {
+			return mappingStructure.get(key.asUserExtSource());
+		}
+
+		for (Attribute attribute : key.getAttributes()) {
+
+			//logins are either single value string, or multi value list
+			List<String> loginsFromAttribute;
+			if (attribute.getType().equals(String.class.getName())) {
+				loginsFromAttribute = Arrays.asList(attribute.valueAsString());
+			} else {
+				loginsFromAttribute = attribute.valueAsList();
+			}
+
+			//check if extSource with login from ues attribute is in the map
+			for (UserExtSource ues : mappingStructure.keySet()) {
+				if (ues.getExtSource().equals(key.asUserExtSource().getExtSource())) {
+					try {
+						Attribute attributeToCompare = perunBl.getAttributesManagerBl().getAttribute(sess, ues, attribute.getName());
+
+						List<String> attributeToCompareValue;
+						if (attributeToCompare.getType().equals(String.class.getName())) {
+							attributeToCompareValue = Arrays.asList(attribute.valueAsString());
+						} else {
+							attributeToCompareValue = attribute.valueAsList();
+						}
+
+						//calculate intersection of attribute values
+						List<String> result = attributeToCompareValue.stream().distinct().filter(loginsFromAttribute::contains).collect(Collectors.toList());
+
+						if (!result.isEmpty()) {
+							return mappingStructure.get(ues);
+						}
+					} catch (WrongAttributeAssignmentException | AttributeNotExistsException e) {
+						//attribute not found, just continue
+					}
+				}
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Updates (or creates) ues attributes to additional userExtSources of candidate.
+	 *
+	 * @param sess perun session
+	 * @param candidateSync candidate whose ues attributes are being updated
+	 */
+	private void updateAdditionalUserExtSourceAttributes(PerunSession sess, CandidateSync candidateSync) {
+		for (RichUserExtSource ues : candidateSync.getRichUserExtSources()) {
+			for (Attribute attribute : ues.getAttributes()) {
+				try {
+					Attribute attributeToAdd = perunBl.getAttributesManagerBl().getAttribute(sess, ues.asUserExtSource(), attribute.getName());
+
+					if (attributeToAdd.getValue() == null || !attributeToAdd.getValue().equals(attribute.getValue())) {
+						perunBl.getAttributesManagerBl().setAttribute(sess, ues.asUserExtSource(), attribute);
+					}
+				} catch (WrongAttributeAssignmentException | WrongAttributeValueException | WrongReferenceAttributeValueException e) {
+					log.error("Attribute {} for UserExtSource {} was not updated due to exception: {}", attribute, ues.asUserExtSource(), e);
+				} catch (AttributeNotExistsException e) {
+					try {
+						perunBl.getAttributesManagerBl().setAttribute(sess, ues.asUserExtSource(), attribute);
```

Comment:
Yes, because this means that definition does not exist in Perun. In such a case you can create a new user ext source attribute definition and then set the value again.